File: client/src/types/brokerChat.ts
// Shared types for the broker↔company chat. Both the broker surface
// (pages/broker/BrokerMessages) and the company surface
// (pages/app/broker-chat/BrokerChat) render the same shapes.

export type ChatReferenceType =
  | 'claim'
  | 'loss_run'
  | 'document'
  | 'flagged_data'
  | 'incident'
  | 'submission'
  | 'policy'
  | 'general'

export type ChatSide = 'broker' | 'company'

export interface MessageReference {
  type: ChatReferenceType
  id?: string | null
  label: string
}

export interface ChatMessage {
  id: string
  conversation_id: string
  sender_user_id: string
  sender_side: ChatSide
  sender_name: string
  body: string
  reference?: MessageReference | null
  client_message_id?: string | null
  created_at: string
  edited_at?: string | null
}

export interface Conversation {
  id: string
  broker_id: string
  company_id: string
  company_name?: string | null
  broker_name?: string | null
  subject?: string | null
  status: 'open' | 'archived'
  reference?: MessageReference | null
  created_by_side: ChatSide
  last_message_at?: string | null
  last_message_preview?: string | null
  unread_count: number
  created_at: string
}

export interface ConversationListResponse {
  conversations: Conversation[]
  total_unread: number
}

export interface ChatTarget {
  id: string
  name: string
}

export interface BrokerChatSummary {
  has_active_broker: boolean
  unread: number
  brokers: ChatTarget[]
}

// Request bodies
export interface SendMessageBody {
  body: string
  reference?: MessageReference | null
  client_message_id?: string | null
}

// Side-agnostic create payload used by the shared component; the per-side
// adapter maps `targetId` to company_id (broker) or broker_id (company).
export interface CreateConversationInput {
  targetId?: string | null
  subject?: string | null
  reference?: MessageReference | null
  body?: string | null
}

File: server/app/matcha/models/broker_chat.py
"""Request/response shapes for the broker↔company chat.

The same conversation/message rows are surfaced to two audiences — a broker
member (via ``routes/broker/chat.py``) and a company user (via
``routes/broker_chat_company.py``) — so the response models are side-agnostic
and carry ``sender_side`` / ``created_by_side`` for the client to render "us"
vs "them".
"""
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

# Reference kinds a conversation/message may anchor to. Kept as a permissive
# Literal so a message can point at any shared record the two parties discuss.
ReferenceType = Literal[
    "claim",
    "loss_run",
    "document",
    "flagged_data",
    "incident",
    "submission",
    "policy",
    "general",
]

Side = Literal["broker", "company"]


class MessageReference(BaseModel):
    """A pointer to a shared record under discussion (a claim, doc, incident…)."""
    type: ReferenceType
    id: Optional[UUID] = None
    label: str = Field(..., min_length=1, max_length=200)


class ConversationCreate(BaseModel):
    # Broker side supplies company_id (which client to talk to). Company side
    # supplies broker_id only when the company is linked to more than one broker.
    company_id: Optional[UUID] = None
    broker_id: Optional[UUID] = None
    subject: Optional[str] = Field(default=None, max_length=200)
    reference: Optional[MessageReference] = None
    # Optional opening message so the thread isn't born empty.
    body: Optional[str] = Field(default=None, max_length=8000)


class MessageCreate(BaseModel):
    body: str = Field(..., min_length=1, max_length=8000)
    reference: Optional[MessageReference] = None
    # Idempotency token — a retried send with the same value returns the
    # original message instead of duplicating it.
    client_message_id: Optional[UUID] = None


class MessageEdit(BaseModel):
    body: str = Field(..., min_length=1, max_length=8000)


class MarkReadRequest(BaseModel):
    # Newest message the caller has seen. Omit to mark the whole thread read.
    last_read_message_id: Optional[UUID] = None


class MessageOut(BaseModel):
    id: UUID
    conversation_id: UUID
    sender_user_id: UUID
    sender_side: Side
    sender_name: str
    body: str
    reference: Optional[MessageReference] = None
    client_message_id: Optional[UUID] = None
    created_at: datetime
    edited_at: Optional[datetime] = None


class ConversationOut(BaseModel):
    id: UUID
    broker_id: UUID
    company_id: UUID
    company_name: Optional[str] = None
    broker_name: Optional[str] = None
    subject: Optional[str] = None
    status: Literal["open", "archived"]
    reference: Optional[MessageReference] = None
    created_by_side: Side
    last_message_at: Optional[datetime] = None
    last_message_preview: Optional[str] = None
    unread_count: int = 0
    created_at: datetime


class ConversationListOut(BaseModel):
    conversations: list[ConversationOut]
    total_unread: int = 0
